tools/kind-d8.sh

```diff
@@ -0,0 +1,502 @@
+    global:
+      modules:
+        publicDomainTemplate: "%s-127-0-0-1.nip.io"
+        https:
+          mode: Disabled
+    operatorPrometheusCrdEnabled: true
+    operatorPrometheusEnabled: true
+    prometheusCrdEnabled: true
+    prometheusEnabled: true
+    prometheus:
+      longtermRetentionDays: 0
+    monitoringKubernetesControlPlaneEnabled: true
+    ingressNginxEnabled: true
+EOF
+
+  if [[ -n "$D8_LICENSE_KEY" ]]; then
+    generate_ee_access_string "$D8_LICENSE_KEY"
+    cat <<EOF >>${CONFIG_DIR}/config.yml
+  imagesRepo: $D8_REGISTRY_PATH
+  registryDockerCfg: $D8_EE_ACCESS_STRING
+EOF
+  fi
+
+  echo "Creating Deckhouse resource file (${CONFIG_DIR}/resources.yml)..."
+  cat <<EOF >${CONFIG_DIR}/resources.yml
+apiVersion: deckhouse.io/v1
+kind: IngressNginxController
+metadata:
+  name: nginx
+spec:
+  ingressClass: nginx
+  inlet: HostPort
+  hostPort:
+    httpPort: 80
+    httpsPort: 443
+EOF
+}
+
+cluster_create() {
+
+  ${KIND_PATH} create cluster --name "${KIND_CLUSTER_NAME}" --image "${KIND_IMAGE}" --config "${CONFIG_DIR}/kind.cfg"
+
+  if [ "$?" -ne "0" ]; then
+    printf "
+Error creating cluster. If error is like '...port is already allocated' then you need to free ports 80 and 443.
+E.g., you can find programs that use these ports using the following command:
+
+    sudo lsof -n -i TCP@0.0.0.0:80,443 -s TCP:LISTEN
+
+"
+    exit 1
+  fi
+
+  ${KIND_PATH} get kubeconfig --internal --name "${KIND_CLUSTER_NAME}" >${CONFIG_DIR}/kubeconfig
+
+}
+
+deckhouse_install() {
+  echo "Running Deckhouse installation..."
+
+  docker run --pull=always --rm --network kind -v "${CONFIG_DIR}/config.yml:/config.yml" -v "${CONFIG_DIR}/resources.yml:/resources.yml" \
+    -v "${CONFIG_DIR}/kubeconfig:/kubeconfig" ${D8_REGISTRY_PATH}/install:$D8_UPDATE_CHANNEL \
+    bash -c "dhctl bootstrap-phase install-deckhouse --kubeconfig=/kubeconfig --kubeconfig-context=kind-${KIND_CLUSTER_NAME} --config=/config.yml && \
+             dhctl bootstrap-phase create-resources --kubeconfig=/kubeconfig --kubeconfig-context=kind-${KIND_CLUSTER_NAME} --resources=/resources.yml"
+
+  if [ "$?" -ne "0" ]; then
+    echo "Error installing Deckhouse!"
+    exit 1
+  fi
+}
+
+ingress_check() {
+  local retries_max=100
+  local retries_count=0
+
+  retries_count=0
+  echo -n "Waiting for the Ingress controller to be ready..."
+
+  while true; do
+    ingress_ready_pods=$(${KUBECTL_PATH} --context kind-"${KIND_CLUSTER_NAME}" -n d8-ingress-nginx get daemonset/controller-nginx -o jsonpath="{$.status.numberReady}" 2>/dev/null)
+    if [ "$?" -ne "0" ]; then
+      echo -n "."
+      ((retries_count++))
+      sleep 10
+    else
+      # Check Ingress readiness
+      if [ "$ingress_ready_pods" -eq "1" ]; then
+        break
+      else
+        echo -n "."
+        ((retries_count++))
+        sleep 10
+      fi
+    fi
+
+    if [ "$retries_count" -ge "$retries_max" ]; then
+      echo
+      echo "Timeout waiting for the creation of Ingress controller!"
```

Comment:
It would be best to output the last error here, or maybe the whole output of the kubectl -n d8-ingress-nginx get ds -o wide command output.

Reply:
Fixed.
Added more information. Here is an example of the output.

```
Waiting for the Ingress controller to be ready....
Timeout waiting for the creation of Ingress controller!

Here is the output of the 'kubectl -n d8-ingress-nginx get ds controller-nginx -o wide' command:
NAME               DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE   CONTAINERS                                     IMAGES                                                                                                                                                                                                                                                                                                                          SELECTOR
controller-nginx   1         1         1       1            1           <none>          34m   controller,protobuf-exporter,kube-rbac-proxy   registry.deckhouse.io/deckhouse/ce:191932fcff3a7cab75b79c08913f175456d5908c326df73c567f0b16-1646204184061,registry.deckhouse.io/deckhouse/ce:6de8c6e5b8ac80a87f2574a4f0ebcb62df4ce2542434a39318a013f6-1633631210461,registry.deckhouse.io/deckhouse/ce:a4506c2aa962611cf1858c774129d2a4f233502ecc376929aa97b9f5-1639403210069   app=controller,name=nginx

Here is the output of the 'kubectl -n d8-ingress-nginx get po -l app=controller' command:
NAME                     READY   STATUS    RESTARTS   AGE
controller-nginx-k2l6d   3/3     Running   0          34m

If the controller-nginx Pod is in the ContainerCreating status, you most likely have a slow connection. If so, wait a little longer until the controller-nginx Pod becomes Ready. After that, run the following command to get the admin password for Grafana: 'kubectl --context kind-d8 -n d8-system get cm deckhouse -o jsonpath="{.data.prometheus}" | grep password' (in case of success Grafana will be available at http://grafana-127-0-0-1.nip.io/).
```